Structured responses from transcoder

j0sh opened the pull request:
* [x] Structural support for returning values from the transcoder
* [x] Serialize returned value over protobufs, send response
* [x] Fill in transcoder response values (URL, hash, sig, etc)
* [x] Implement response handling on broadcaster
* [x] Monitoring updates
* [x] Download segments from response
* [x] Update playlists
* [ ] Check response signature from transcoder
* [x] Move protobufs into `net` to facilitate access from `core`

j0sh committed:
transcoder: Return a value after transcoding.

j0sh committed:
rpc: Generate code for transcoder responses.

j0sh committed:
rpc: Send transcoded results back to client.

j0sh commented:
Note the structure of the data that we return to the client. We have a choice to make: do we need the hashes of the individual segments, or is it enough to just return the URLs and check the sig of the concatenated hashes? Currently, we return both the individual hashes and the concatenated hashes.

The current protocol also does not allow for this, but I wonder if we should make future allowances for 'partial' success in transcoding, eg only returning a subset of the requested profiles. This would make transcoding less all-or-nothing, and allow for some interesting constructs like 'streaming' responses. To support this, we would need to switch from `list[Segment]` to `map[StreamID]Segment` in that case.

https://github.com/livepeer/go-livepeer/blob/04152802ee86fda59de3af2b02e3dd62f0212b93/server/lp_rpc.proto#L69-L87

j0sh committed:
transcoder: Simplify the internal API.

j0sh committed:
claimmanager: Return transcoded data hash.

j0sh committed:
monitor: Accommodate new transcode response format.

j0sh committed:
broadcaster: Download segments from the transcoder.

j0sh committed:
proto: Move protobufs definitions into net.

Allows us to use the data types outside the server package.

j0sh committed:
proto: Update references to new proto location.

j0sh commented:
> Note the structure of the data that we return to the client. We have a choice to make: do we need the hashes of the individual segments, or is it enough to just return the URLs and check the sig of the concatenated hashes? Currently, we return both the individual hashes and the concatenated hashes.

Reversed course here -- now we only return URLS and a signature. This was mostly out of convenience due to how I chose to return the concatenated hashes: from the claim manager, which already computes those hashes. https://github.com/livepeer/go-livepeer/pull/508/commits/4836db7f57ffd562d19faed2b94fc480080721a4 To return each individual hash, we would need to either refactor the claimmanager hashing procedure, or re-hash everything outside the claimmanager. Path of least resistance, but we can certainly change it up if folks think otherwise.

j0sh committed:
core: Fix tests.

j0sh committed:
server: Fix tests.

j0sh committed:
eth: Fix tests.

ericxtang commented on core/transcoder.go:
```diff
@@ -131,32 +146,30 @@ func (n *LivepeerNode) getSegmentChan(job *ethTypes.Job) (SegmentChan, error) {
+		return nil, err
 	}
-	segChan := &SegChanData{seg: ss, res: make(chan error, 1)}
+	segChan := &SegChanData{seg: ss, res: make(chan *TranscodeResult, 1)}
```
Small thing - `segChan` seems like a misnomer here.  Should be `segChanData`?

j0sh replied:
Fixed https://github.com/livepeer/go-livepeer/pull/512/commits/25e7e8b85b9038aee67a242f037d9794cf26d9d0

ericxtang commented on core/transcoder.go:
```diff
@@ -131,32 +146,30 @@ func (n *LivepeerNode) getSegmentChan(job *ethTypes.Job) (SegmentChan, error) {
 }
 
-func (n *LivepeerNode) transcodeAndBroadcastSeg(seg *stream.HLSSegment, sig []byte, cm eth.ClaimManager, t transcoder.Transcoder, resultStrmIDs []StreamID, config net.TranscodeConfig) error {
+func (n *LivepeerNode) transcodeAndBroadcastSeg(config transcodeConfig, ss *SignedSegment) *TranscodeResult {
```
Even though we are putting the transcoded video into the `VideoCache`, we are no longer "broadcasting" here.  Should we rename this to `transcodeAndCacheSeg`?

j0sh replied:
Fixed https://github.com/livepeer/go-livepeer/pull/512/commits/25e7e8b85b9038aee67a242f037d9794cf26d9d0

ericxtang commented on server/rpc.go:
```diff
@@ -177,7 +178,7 @@ func verifySegCreds(job *lpTypes.Job, segCreds string) (*SegData, error) {
 }
 
-func GetTranscoder(context context.Context, orch Orchestrator, req *TranscoderRequest) (*TranscoderInfo, error) {
+func GetTranscoder(context context.Context, orch Orchestrator, req *net.TranscoderRequest) (*net.TranscoderInfo, error) {
```
Seems `GetTranscoder` is only used in `lphttp.GetTranscoder`.  Make it a private method?

j0sh replied:
Fixed https://github.com/livepeer/go-livepeer/pull/512/commits/276c2c2ee4f2e58760cd458bf1670ab1cd32eb4a

ericxtang reviewed (commented):
Can we also remove the `net.VideoNetwork` from `VideoCache`?  That way it's just a pure cache, and not a weird "network-connected cache" anymore.

j0sh committed:
rpc: Privatize getTranscoder.

j0sh committed:
transcoder: Cosmetic renames.

j0sh committed:
videocache: Flatten VideoNetwork.

This also leaves VideoNetwork with no significant purpose, so fold
the remaining into VideoCache and remove VideoNetwork.

j0sh committed:
core: Additional test cleanups.

j0sh committed:
transcoder: Move GetClaimManager into transcoder.go.

Also document the logical layout of the code within LivepeerNode.

j0sh committed:
orchestrator: Rename file.

j0sh committed:
videosource: Rename from VideoCache.

j0sh committed:
rpc: Fix crash on non-200 response.

j0sh committed:
core: Remove the notion of a NodeID.

Also add some tests around StreamID / ManifestID.

j0sh committed:
livepeer: Switch default ports around.

HTTP/RPC becomes 8935
CLI becomes 7935

j0sh commented:
VideoNetwork was removed from VideoCache in https://github.com/livepeer/go-livepeer/pull/512/commits/2da29308070528f6f4c151292f4b264d26e0cfd3

j0sh committed:
livepeer: Use blockchain ServiceURI rather than locally inferred URI.

darkdarkdragon committed:
Using blocking connection to the rpc server - this will give us clear error in case transcoder is down

j0sh committed:
core: Length checks for ManifestID/StreamID.

Prevents crashes on malformed input.

j0sh committed:
jobwatcher: Make log parsing more robust.

Prevents crashes with malformed StreamIDs.

j0sh committed:
jobwatcher: Retry the assigned transcoder lookup.

More robust against sporadic Eth RPC failures.

j0sh committed:
rpc: Misc fixes.

j0sh committed:
README: Update.

ericxtang committed:
Merge pull request #513 from livepeer/ja/nodeid

Remove Node ID

yondonfu committed:
Merge pull request #516 from livepeer/it/errors_logging

Better errors reporting

yondonfu committed:
Merge pull request #512 from livepeer/ja/v2followup

Follow up changes to v2 feedback

yondonfu committed:
Merge pull request #511 from livepeer/ja/fixtests

Fix tests

yondonfu reviewed (approved):
🚢 Tracking the remaining todo here #523
